lib/markdown_videos/services/service_base.rb

```diff
@@ -0,0 +1,124 @@
+module MarkdownVideos
+  module Services
+
+    class ServiceBase
+
+      attr_accessor :id,
+                    :alt_text,
+                    :markdown_url,
+                    :width,
+                    :height,
+                    :wrapper,
+                    :classname
+
+      def self.support(markdown_url)
+        markdown_url =~ regexp
+      end
+
+      def self.regexp
+        raise "#{self.class} class should respond to `regexp`"
+      end
+
+
+      def initialize(alt_text, markdown_url, options)
+        @alt_text = alt_text
+        @markdown_url = markdown_url
+        options.each { |option, value| self.send("#{option}=", value) }
+      end
+
+      # URL to be consumed by markup
+      def url
+        raise "#{self.class} instance should respond to `url`"
+      end
+
+      # List of authorized url parameters
+      #
+      # @return [Array<Symbol>] e.g.: [:start]
+      def url_parameters
+        []
+      end
+
+      def width
+        @width || 500
+      end
+
+      def height
+        @height || 281
+      end
+
+      # @return [String] escaped alternative text given through markdown syntax
+      def alt_text
+        CGI::escapeHTML(@alt_text)
+      end
+
+      # @return [String] video id, usually consumed by `url` method in service class
+      def id
+        @id ||= self.class.regexp.match(markdown_url)[2]
+      end
+
+
+      # @return [String] service URL to be consumed by markup
+      def service_url
+        given_url = Addressable::URI.parse(markdown_url)
+        given_url_parameters = given_url.query_values || {}
+        given_url_parameters.select! { |attribute, value| url_parameters.include?(attribute.to_sym) && (!value.nil? || !value.empty?) }
+
+        service_url = Addressable::URI.parse(url)
+        service_url.query_values = given_url_parameters unless given_url_parameters.nil? || given_url_parameters.empty?
```

Comment:
Your code is good, just another idea here. I feel like it's a tiny bit easier to read because you dont need the temp var `given_url_parameters `.

```
given_url.query_values ||= {}
valid_parameters = ->(param, value) { (!value.nil? || !value.empty?) && url_parameters.include?(param.to_sym) }
given_url.query_values.select! &:valid_parameters
```